app/naive_bayes.py

```diff
@@ -0,0 +1,286 @@
+                    dict_link_features["Host_Architecture"]
+                ]
+                matrix_link_data.append(link_row)
+        return matrix_link_data
+
+    def _get_useful_facts(self, cur_link : Link) -> Dict[str, str]:
+        """
+        Selects generalizable facts from link object for data querying
+        """
+        useful_facts = {}
+        if(len(cur_link.used) > 0):
+            for used_fact in cur_link.used:
+                # exclude fact types that are unique to a specific host
+                if str(used_fact.trait).startswith(("host.", "remote.", "file.last.", "domain.user.")) == False:
+                    useful_facts[str(used_fact.trait)] = str(used_fact.value)
+        return useful_facts
+
+    def _query_link_matrix(self, matrix_link_data : List[List], query_features : Dict) -> List[List]:
+        """
+        Query link matrix for links containing specific features and values and 
+        return new matrix composed of column names row and the link rows containing right feat and val
+
+        :param matrix_link_data: link matrix that will be queried
+        :param query_features: dict of {feature_names: values} that matrix_link_data will be queried by
+        """
+        col_name_to_col_index = {}
+        list_col_names = matrix_link_data[0]
+        for index_feature_names in range(len(list_col_names)):
+            col_name_to_col_index[list_col_names[index_feature_names]] = index_feature_names
+
+        queried_link_matrix = [list_col_names]
+        for row_index in range(1, len(matrix_link_data)):
+            cur_link = matrix_link_data[row_index]
+            row_passes_conditions = True
+            for feat_name, feat_value in query_features.items():
+                feat_index = col_name_to_col_index[feat_name]
+                if feat_name == "Link_Facts":
+                    cur_facts_dict = cur_link[feat_index]
+                    for req_fact_type, req_fact_val in query_features["Link_Facts"].items():
+                        # check that current link contains required fact type and required fact value
+                        if req_fact_type not in cur_facts_dict or str(cur_facts_dict[req_fact_type]) != str(req_fact_val):
+                            row_passes_conditions = False
+                else:
+                    if str(cur_link[feat_index]) != str(feat_value):
+                        row_passes_conditions = False
+            if row_passes_conditions:
+                queried_link_matrix.append(cur_link)
+        return queried_link_matrix
+
+    def _get_link_success_probability(self, query_features : Dict) -> Union[float, None]: 
+        """
+        Calculates Naive Bayes probability for a link (with features and values from query_features) executing successfully.
+        Returns success probability, and number of link observations in dataset.
+        """
+        num_total_links = len(self.matrix_past_links)-1
```

Comment:
Again here, if you take the feature names out of `matrix_past_links` you wouldn't need this -1 any more.